govcd/metadata_v2.go

```diff
@@ -848,23 +868,149 @@ func mergeMetadataAndWait(client *Client, requestUri string, metadata map[string
+	ObjectName *string        // Name of the object
+	KeyRegex   *regexp.Regexp // A regular expression to filter out metadata keys
+	ValueRegex *regexp.Regexp // A regular expression to filter out metadata values
+}
+
+// filterMetadata filters all metadata entries, given a slice of metadata that needs to be ignored. It doesn't
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterMetadata(allMetadata *types.Metadata, href, objectName string, metadataToIgnore []IgnoredMetadata) (*types.Metadata, error) {
+	if len(metadataToIgnore) == 0 {
+		return allMetadata, nil
+	}
+
+	result := &types.Metadata{
+		XMLName:       allMetadata.XMLName,
+		Xmlns:         allMetadata.Xmlns,
+		HREF:          allMetadata.HREF,
+		Type:          allMetadata.Type,
+		Xsi:           allMetadata.Xsi,
+		Link:          allMetadata.Link,
+		MetadataEntry: nil,
+	}
+
+	var filteredMetadata []*types.MetadataEntry
+	for _, originalEntry := range allMetadata.MetadataEntry {
+		_, err := filterSingleMetadataEntry(originalEntry.Key, href, objectName, &types.MetadataValue{Domain: originalEntry.Domain, TypedValue: originalEntry.TypedValue}, metadataToIgnore)
+		if err == nil || !strings.Contains(err.Error(), "ignored") {
+			filteredMetadata = append(filteredMetadata, originalEntry)
+		}
+	}
+	result.MetadataEntry = filteredMetadata
+	return result, nil
+}
+
+// filterSingleMetadataEntry filters a single metadata entry given a slice of metadata that needs to be ignored. It doesn't
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterSingleMetadataEntry(key, href, objectName string, metadataEntry *types.MetadataValue, metadataToIgnore []IgnoredMetadata) (*types.MetadataValue, error) {
+	if len(metadataToIgnore) == 0 {
+		return metadataEntry, nil
+	}
+
+	objectType, err := getMetadataObjectTypeFromHref(href)
+	if err != nil {
+		return nil, err
+	}
+	for _, entryToIgnore := range metadataToIgnore {
+		if entryToIgnore.ObjectType == nil && entryToIgnore.ObjectName == nil && entryToIgnore.KeyRegex == nil && entryToIgnore.ValueRegex == nil {
+			continue
+		}
+
+		if (entryToIgnore.ObjectType == nil || strings.TrimSpace(*entryToIgnore.ObjectType) == "" || *entryToIgnore.ObjectType == objectType) &&
+			(entryToIgnore.ObjectName == nil || strings.TrimSpace(*entryToIgnore.ObjectName) == "" || strings.TrimSpace(objectName) == "" || *entryToIgnore.ObjectName == objectName) &&
+			(entryToIgnore.KeyRegex == nil || entryToIgnore.KeyRegex.MatchString(key)) &&
+			(entryToIgnore.ValueRegex == nil || entryToIgnore.ValueRegex.MatchString(metadataEntry.TypedValue.Value)) {
+			return nil, fmt.Errorf("the entry with key '%s' and value '%v' is being ignored", key, metadataEntry.TypedValue.Value)
+		}
+	}
+	return metadataEntry, nil
+}
```

Comment:
One more thought - should we add some verbosity logging here so that we can easier troubleshoot if there are bug reports. In my mind I'd have this - what filter came in and did it match or not a particular entry?

To dump the filter itself nicely you could add `IgnoredMetadata.String()` method which will satisfy `stringer` interface and `fmt.X` will automatically print it using this method

Reply:
That's a good idea, done.
Let me know your thoughts.